File: mobile-app/test/features/swipe_undo/swipe_undo_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:projectgamma_mobile/features/swipe_undo/swipe_undo.dart';

void main() {
  group('inverseMutation', () {
    test('TOGGLE_TASK ist idempotent (selbe payload)', () {
      final inv = inverseMutation('TOGGLE_TASK', {'projectId': 'p1', 'taskId': 't1'});
      expect(inv?.type, 'TOGGLE_TASK');
      expect(inv?.payload, {'projectId': 'p1', 'taskId': 't1'});
    });

    test('DISMISS_IDEA wird zu REACTIVATE_IDEA', () {
      final inv = inverseMutation('DISMISS_IDEA', {'projectId': 'p1', 'ideaId': 'i9'});
      expect(inv?.type, 'REACTIVATE_IDEA');
      expect(inv?.payload['ideaId'], 'i9');
    });

    test('CONVERT_IDEA wird zu REACTIVATE_IDEA', () {
      final inv = inverseMutation('CONVERT_IDEA', {'projectId': 'p1', 'ideaId': 'i9'});
      expect(inv?.type, 'REACTIVATE_IDEA');
    });

    test('REACTIVATE_IDEA wird zu DISMISS_IDEA', () {
      final inv = inverseMutation('REACTIVATE_IDEA', {'projectId': 'p1', 'ideaId': 'i9'});
      expect(inv?.type, 'DISMISS_IDEA');
    });

    test('unbekannter Typ → null', () {
      expect(inverseMutation('FOO_BAR', const {}), isNull);
    });

    test('payload-kopie ist defensiv (kein alias)', () {
      final original = {'projectId': 'p1', 'taskId': 't1'};
      final inv = inverseMutation('TOGGLE_TASK', original)!;
      inv.payload['taskId'] = 'changed';
      expect(original['taskId'], 't1');
    });
  });

  group('UndoStack', () {
    UndoEntry mkEntry(String id, DateTime t) => UndoEntry(
          id: id,
          label: 'x',
          mutationType: 'TOGGLE_TASK',
          payload: const {},
          createdAt: t,
        );

    test('push + consume gibt eintrag, danach null', () {
      final s = UndoStack();
      s.push(mkEntry('a', DateTime.now()));
      expect(s.consume('a')?.id, 'a');
      expect(s.consume('a'), isNull);
    });

    test('active filtert abgelaufene', () {
      var t = DateTime.utc(2026, 5, 10, 12, 0, 0);
      final s = UndoStack(ttl: const Duration(seconds: 5), now: () => t);
      s.push(mkEntry('old', t.subtract(const Duration(seconds: 10))));
      s.push(mkEntry('new', t));
      expect(s.active.map((e) => e.id), ['new']);
    });

    test('purgeExpired entfernt physisch', () {
      var t = DateTime.utc(2026, 5, 10, 12, 0, 0);
      final s = UndoStack(ttl: const Duration(seconds: 5), now: () => t);
      s.push(mkEntry('old', t.subtract(const Duration(seconds: 10))));
      s.push(mkEntry('new', t));
      expect(s.purgeExpired(), 1);
      expect(s.size, 1);
    });

    test('UndoEntry.inverse delegiert an inverseMutation', () {
      final e = UndoEntry(
        id: 'x', label: 'l', mutationType: 'DISMISS_IDEA',
        payload: {'ideaId': 'i'}, createdAt: DateTime.now(),
      );
      expect(e.inverse?.type, 'REACTIVATE_IDEA');
    });
  });
}

File: release/ProjectGamma/desktop-app/swipe_undo.test.js
'use strict';

const { test } = require('node:test');
const assert = require('node:assert');
const { inverseMutation, UndoStack, isSwipeable } = require('./swipe_undo');

test('inverseMutation: TOGGLE_TASK ist idempotent', () => {
  const r = inverseMutation('TOGGLE_TASK', { projectId: 'p', taskId: 't1' });
  assert.deepStrictEqual(r, {
    type: 'TOGGLE_TASK',
    payload: { projectId: 'p', taskId: 't1' },
  });
});

test('inverseMutation: DISMISS_IDEA → REACTIVATE_IDEA', () => {
  const r = inverseMutation('DISMISS_IDEA', { projectId: 'p', ideaId: 'i1' });
  assert.strictEqual(r.type, 'REACTIVATE_IDEA');
  assert.deepStrictEqual(r.payload, { projectId: 'p', ideaId: 'i1' });
});

test('inverseMutation: CONVERT_IDEA → REACTIVATE_IDEA', () => {
  const r = inverseMutation('CONVERT_IDEA', { ideaId: 'x' });
  assert.strictEqual(r.type, 'REACTIVATE_IDEA');
});

test('inverseMutation: REACTIVATE_IDEA → DISMISS_IDEA', () => {
  const r = inverseMutation('REACTIVATE_IDEA', { ideaId: 'x' });
  assert.strictEqual(r.type, 'DISMISS_IDEA');
});

test('inverseMutation: unbekannter Typ → null', () => {
  assert.strictEqual(inverseMutation('FOO_BAR', {}), null);
});

test('inverseMutation: payload wird kopiert (kein shared ref)', () => {
  const src = { projectId: 'p', taskId: 't' };
  const r = inverseMutation('TOGGLE_TASK', src);
  r.payload.taskId = 'changed';
  assert.strictEqual(src.taskId, 't');
});

test('UndoStack: push + consume', () => {
  const s = new UndoStack({ ttlMs: 1000, now: () => 1000 });
  s.push({ id: 'a', mutationType: 'TOGGLE_TASK', payload: { taskId: '1' } });
  assert.strictEqual(s.size, 1);
  const popped = s.consume('a');
  assert.strictEqual(popped.id, 'a');
  assert.strictEqual(s.size, 0);
  assert.strictEqual(s.consume('a'), null);
});

test('UndoStack: active filtert abgelaufene', () => {
  let now = 0;
  const s = new UndoStack({ ttlMs: 100, now: () => now });
  s.push({ id: 'old', mutationType: 'TOGGLE_TASK', payload: {} });
  now = 50;
  s.push({ id: 'new', mutationType: 'TOGGLE_TASK', payload: {} });
  now = 200;
  const active = s.active;
  assert.strictEqual(active.length, 0);
  now = 120;
  assert.strictEqual(s.active.length, 1);
  assert.strictEqual(s.active[0].id, 'new');
});

test('UndoStack: purgeExpired entfernt nur abgelaufene', () => {
  let now = 0;
  const s = new UndoStack({ ttlMs: 100, now: () => now });
  s.push({ id: 'old', mutationType: 'TOGGLE_TASK', payload: {} });
  now = 50;
  s.push({ id: 'new', mutationType: 'TOGGLE_TASK', payload: {} });
  now = 120;
  const removed = s.purgeExpired();
  assert.strictEqual(removed, 1);
  assert.strictEqual(s.size, 1);
  assert.strictEqual(s._entries[0].id, 'new');
});

test('UndoStack: push ohne id wirft', () => {
  const s = new UndoStack();
  assert.throws(() => s.push({ mutationType: 'TOGGLE_TASK' }));
});

test('isSwipeable: ghost-pending items sind nicht swipe-bar', () => {
  assert.strictEqual(isSwipeable({ id: '1' }), true);
  assert.strictEqual(isSwipeable({ id: '1', __pending: true }), false);
  assert.strictEqual(isSwipeable(null), false);
});
